### ISSUE TITLE
customized elpy-rpc-python-command is not respected

### ISSUE THREAD

**Issue by landisrm** [jorgenschaefer/elpy#201]
I'm really looking forward to trying out elpy after trying out many other emacs setups to use python.  Unfortunately, I keep getting an _Elpy installation_ buffer opening, which says: "Elpy Installation Instructions  The elpy Python module was not found".  I think this is because Elpy is insisting on using the wrong version of python on my system.

For background, I'm running this on a CentOS machine where I have two different python versions running.  The system version is:

```
$ which python
/usr/bin/python
```

But I've installed python2.7 that I use for everything besides system functions. 

```
$ which python2.7
/usr/local/bin/python2.7
```

I've installed elpy for this version.  I don't know much about virtualenv and I don't use it.

I'm using CentOS 6.5, Emacs 24.2.1, Python 2.7.3 (at least, I wish I was), IPython 1.2.1, elpy 1.2.1

So, here's what I've tried to figure this issue out:
I can successfully open python2.7 and import elpy.

I set `elpy-rpc-python-command` using the customize variable buffer so that it is set in my .emacs file.  `M-x describe-variable RET elpy-rpc-python-command` returns /usr/local/bin/python2.7, as it should.

However, when I C-c C-z, the buffer I get is  /usr/bin/python and of course it can't find elpy (the _elpy-rpc_ buffer says: /usr/bin/python: No module named elpy)

I've edited elpy.el manually to set `elpy-rpc-python-command` to /usr/local/bin/python2.7, but C-c C-z still starts /usr/bin/python.

Oddly, after trying this latter command, I just noticed a buffer _Elpy Version Mismatch_ informing me that Elpy Emacs Lisp version is 1.2.1 but Elpy Python Version is 1.3.0.  So, now that I've changed elpy.el to use /usr/local/bin/python2.7, it is finding python with elpy in it, but not using it for C-c C-z (and presumably not when elpy is initialized).

Incidentally, I  just fixed the issue by additionally setting `elpy-use-ipython`  in my .emacs, which works because my ipython uses python2.7.  (I also downgraded the version of elpy installed with pip to 1.2.1).

To sum up, elpy seems to insist on using the default version of python on my system, even after I've told it to use a specific alternative; however that alternative version does seem to be found by part of elpy if I set `elpy-rpc-python-command` manually in elpy.el

Apologies for the length of this note, and for the many conventions which I probably flaunted because I'm not very experienced at reporting bugs on github.

Many thanks again for your work on elpy!

**Comment by spluque** [comment]
Just to be clear, do elpy-rpc-python-command need to match python-shell-interpreter?  I still want to use Ipython for interactive work, but if Elpy needs a non-interactive shell, then I guess I can keep the latter variable as ipython and the former at the standard non-interactive python shell.  Is that right?

**Comment by jorgenschaefer** [comment]
> Is that right?

That's exactly the reason why there are two distinct variables :-)

**Comment by spluque** [comment]
On Thu, 31 Mar 2016 08:21:19 -0700,
Jorgen Schäfer <notifications@github.com> wrote:

>     Is that right?  That's exactly the reason why there are two
> distinct variables :-)

Nice!

**Comment by jorgenschaefer** [comment]
Elpy uses `elpy-rpc-python-command` for the RPC process (e.g. completions, find definitions). Which variable to use to configure the interactive Python process depends on your Emacs version. In recent Emacsen, that variable is called `python-shell-interpreter`. Does this help?

**Comment by Nicholas-Mitchell** [comment]
Thanks for the speedy response!

I have managed to get that working  :smile: 
I had also been using ipython, so have managed to get it to work with this setup.

```lisp
 (setq python-shell-interpreter "/usr/local/bin/ipython"
	  python-shell-interpreter-args "-i --simple-prompt")
```

As a side note - [it is necessary to install Ipython 5.x](https://ipython.readthedocs.io/en/stable/install/kernel_install.html#kernels-for-python-2-and-3) in my case, as the latest version 6 only supports python >=3.3.

I suppose the next step will be to make Emacs ask me which version of python to use every time I start an interactive prompt. Would it be as simple as passing a function to `python-shell-interpreter` that then lets me pick from a (hard-coded?) list of python binaries?

**Comment by jorgenschaefer** [comment]
You will most likely have to write your own command that let-binds the `python-shell-interpreter` variable while calling `run-python` …

**Comment by Nicholas-Mitchell** [comment]
I did manage to get this working as I would like by adding very simple functions to my `init.el`.

@galaunay @jorgenschaefer 
I cannot judge if any other users might find this useful to integrate into Elpy - would just mean creating a new variable to allow users to specify places where to find pythoninterpreters on their machines and then a function to choose which one gets used - see my code below if my explanation didn't help:

```lisp
  (defun nwm/which-python-binary ()
    "Select one of the python binaries to be used in the variable: python-shell-interpreter
     - Selecting 'ipython' will just take the first thing found in PATH.
     - Selecting '/usr/bin/python' will use the Linux system python."
  
    (interactive)
    (let ((binaries '("ipython" "/usr/bin/python/" "/usr/local/bin/ipython2")))
      (message "%s" (ido-completing-read+ "Select Python binary:" binaries))))

  (defun nwm/select-python-binary ()
    "Set the python binary to be used for current and future scripts (i.e. until changed again).
     Choose binary from list, contained in: nwm/which-python-binary"

    (interactive)
    (setq selected-binary (nwm/which-python-binary))
    (setq python-shell-interpreter selected-binary))

```
I suppose the key part is the last line, setting `python-shell-interpreter`. Using `ido` just makes things a little more convenient. As I wrote in issue #1328 - this actually seems to play nicely with toggling dedicated shells too.

I just call the second function whenever I want to be using a different binary. This solves the problem for me that I cannot seem to get all available system binaries to be shown nicely with the available virtual environments the `pyvenv-workon` offers. So I keep that side simple:

```lisp
(setenv "WORKON_HOME" (concat (getenv "HOME") "/anaconda3/envs/"))
```

**Comment by galaunay** [comment]
> I would like to set Elpy to use the system Python by default, which is /usr/bin/python.
I generally use Anaconda and more specifically their environments, using pyvenv-workon.

Sorry, but I don't get why you want to use the system python shell interpreter with a virtual env activated.

Could you not just deactivate anaconda virtualenv ?

**Comment by Nicholas-Mitchell** [comment]
That was an unclear explanation - I don't use them together. I want the system python to be their by default - I activate an Anaconda environment when I need something other then system python.
